crates/viewer/re_renderer/src/renderer/video.rs

```diff
@@ -0,0 +1,544 @@
+                    ));
+                }
+            })?;
+
+            let queue = render_context.queue.clone();
+
+            // NOTE: both textures are assumed to be rgba8unorm
+            let texture = super::alloc_video_frame_texture(
+                &render_context.device,
+                &render_context.gpu_resources.textures,
+                data.config.coded_width as u32,
+                data.config.coded_height as u32,
+            );
+            let zeroed_texture_float = GpuTexture2D::new(
+                render_context
+                    .texture_manager_2d
+                    .zeroed_texture_float()
+                    .clone(),
+            )
+            .expect("expected texture to be 2D");
+
+            let mut this = Self {
+                data,
+                queue,
+                texture,
+                zeroed_texture_float,
+
+                decoder,
+
+                frames,
+                last_used_frame_timestamp: TimeMs::new(f64::MAX),
+                current_segment_idx: usize::MAX,
+                current_sample_idx: usize::MAX,
+            };
+
+            // immediately enqueue some frames, assuming playback at start
+            this.reset();
+            let _ = this.get_frame(TimeMs::new(0.0));
+
+            Some(this)
+        }
+
+        pub fn duration_ms(&self) -> f64 {
+            self.data.duration.as_f64()
+        }
+
+        pub fn width(&self) -> u32 {
+            self.data.config.coded_width as u32
+        }
+
+        pub fn height(&self) -> u32 {
+            self.data.config.coded_height as u32
+        }
+
+        pub fn get_frame(&mut self, timestamp: TimeMs) -> GpuTexture2D {
+            if timestamp.as_f64() < 0.0 {
+                // TODO(andreas): This is a hack, we should have a better way to handle this
+                return self.zeroed_texture_float.clone();
+            }
+
+            let Some(segment_idx) =
+                latest_at_idx(&self.data.segments, |segment| segment.timestamp, &timestamp)
+            else {
+                return self.texture.clone();
+            };
+
+            let Some(sample_idx) = latest_at_idx(
+                &self.data.segments[segment_idx].samples,
+                |sample| sample.timestamp,
+                &timestamp,
+            ) else {
+                // segments are never empty
+                unreachable!();
+            };
+
+            if segment_idx != self.current_segment_idx {
+                let segment_distance = segment_idx as isize - self.current_segment_idx as isize;
+                if segment_distance == 1 {
+                    // forward seek to next segment
+                    self.enqueue_all(segment_idx + 1);
+                } else {
+                    // forward seek by N>1 OR backward seek across segments
+                    self.reset();
+                    self.enqueue_all(segment_idx);
+                    self.enqueue_all(segment_idx + 1);
```

Comment:
Yes, the rule here is 2 full segments queued ahead of time, so that we can always transition smoothly to the next segment without waiting. Will document